app/src/main/kotlin/compiler/intermediate_form/FunctionDetailsGenerator.kt

```diff
@@ -3,83 +3,271 @@ package compiler.intermediate_form
+                CFGLinkType.CONDITIONAL_FALSE to functionCfg.conditionalFalseLinks
+            )
+
+            for ((linkType, links) in linksToIterateOver) {
+                for ((from, to) in links) {
+                    if (to !in finalNodes)
+                        modifiedCfgBuilder.addLink(Pair(from, linkType), to)
+                    else {
+                        val newReturnNode = IntermediateFormTreeNode.RegisterWrite(FUNCTION_RESULT_REGISTER, to)
+                        modifiedCfgBuilder.addLink(Pair(from, linkType), newReturnNode)
+                    }
+                }
+            }
+            return FunctionCallIntermediateForm(
+                modifiedCfgBuilder.build(),
+                IntermediateFormTreeNode.RegisterRead(FUNCTION_RESULT_REGISTER)
+            )
+        }
+
+        if (function.returnType == Type.Unit)
+            return FunctionCallIntermediateForm(cfgBuilder.build(), null)
+        return modifyReturnNodesToStoreResultInAppropriateRegister(cfgBuilder.build())
     }
 
-    // add function graph
-    cfgBuilder.addAllFrom(fdg.functionCFG)
-    cfgBuilder.addLink(last, fdg.functionCFG.entryTreeRoot!!)
+    fun genPrologue(): ControlFlowGraph {
+        val cfgBuilder = ControlFlowGraphBuilder()
 
-    fun modifyReturnNodesToStoreResultInAppropriateRegister(functionCfg: ControlFlowGraph): FunctionCallIntermediateForm {
-        val finalNodes = functionCfg.finalTreeRoots
+        // save rbp
+        val pushRbp = IntermediateFormTreeNode.StackPush(IntermediateFormTreeNode.RegisterRead(BASE_POINTER_REGISTER))
+        val movRbpRsp = IntermediateFormTreeNode.RegisterWrite(
+            BASE_POINTER_REGISTER,
+            IntermediateFormTreeNode.RegisterRead(STACK_POINTER_REGISTER)
+        )
 
-        val modifiedCfgBuilder = ControlFlowGraphBuilder()
+        cfgBuilder.addLink(null, pushRbp)
+        cfgBuilder.addLink(Pair(pushRbp, CFGLinkType.UNCONDITIONAL), movRbpRsp)
+        var last: Pair<IFTNode, CFGLinkType> = Pair(movRbpRsp, CFGLinkType.UNCONDITIONAL)
 
-        val linksToIterateOver = hashMapOf(
-            CFGLinkType.UNCONDITIONAL to functionCfg.unconditionalLinks,
-            CFGLinkType.CONDITIONAL_TRUE to functionCfg.conditionalTrueLinks,
-            CFGLinkType.CONDITIONAL_FALSE to functionCfg.conditionalFalseLinks
+        // update display
+        val savePreviousRbp = IntermediateFormTreeNode.StackPush(
+            IntermediateFormTreeNode.MemoryRead(Addressing.Displacement(displayAddress + memoryUnitSize * depth))
         )
+        val updateRbpAtDepth = IntermediateFormTreeNode.MemoryWrite(
+            Addressing.Displacement(displayAddress + memoryUnitSize * depth),
+            IntermediateFormTreeNode.RegisterRead(BASE_POINTER_REGISTER)
+        )
+        cfgBuilder.addLink(last, savePreviousRbp)
+        last = Pair(savePreviousRbp, CFGLinkType.UNCONDITIONAL)
+        cfgBuilder.addLink(last, updateRbpAtDepth)
```

Comment:
This should be a function of `ControlFlowGraphBuilder`. `addLinkToLast` or something similar.